contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -4290,3 +4304,528 @@ sp_reset_connection_internal(PG_FUNCTION_ARGS)
+    docSizeCol->typeName = makeTypeNameFromOid(INT8OID, -1);
+    docSizeCol->is_local = true;
+    columns = lappend(columns, docSizeCol);
+    
+    nsSizeCol = makeNode(ColumnDef);
+    nsSizeCol->colname = "ns_size";
+    nsSizeCol->typeName = makeTypeNameFromOid(INT8OID, -1);
+    nsSizeCol->is_local = true;
+    columns = lappend(columns, nsSizeCol);
+    
+    isNsNullCol = makeNode(ColumnDef);
+    isNsNullCol->colname = "is_namespace_null";
+    isNsNullCol->typeName = makeTypeNameFromOid(BOOLOID, -1);
+    isNsNullCol->is_local = true;
+    columns = lappend(columns, isNsNullCol);
+    
+    xmlDataCol = makeNode(ColumnDef);
+    xmlDataCol->colname = "xml_data";
+    xmlDataCol->typeName = makeTypeNameFromOid(XMLOID, -1);
+    xmlDataCol->is_local = true;
+    columns = lappend(columns, xmlDataCol);
+    
+    nsDataCol = makeNode(ColumnDef);
+    nsDataCol->colname = "ns_data";
+    nsDataCol->typeName = makeTypeNameFromOid(XMLOID, -1);
+    nsDataCol->is_local = true;
+    columns = lappend(columns, nsDataCol);
+    
+    return columns;
+}
+
+/*
+ * create_xml_handle_temp_table:
+ *     Creates a temporary table to store XML document handles and their associated data.
+ *
+ * This function creates a temporary table named "#xml_handle_temp_table" that persists
+ * for the duration of the session. The table stores XML documents and their associated
+ * namespace information for use with the sp_xml_preparedocument and sp_xml_removedocument
+ * procedures. 
+ * 
+ * The function temporarily elevates privileges to "bbf_role_admin" to ensure it has
+ * the necessary permissions to create the table. It also sets the SQL dialect
+ * to TSQL to ensure proper table creation semantics.
+ */
+
+ void 
+ create_xml_handle_temp_table()  
+ {
+     CreateStmt   *stmt = makeNode(CreateStmt);
+     RangeVar     *relation = makeRangeVar(NULL,"#xml_handle_temp_table", -1);
+     Oid          save_userid;
+     int          save_sec_context;
+     int          saved_dialect = sql_dialect;
+     ObjectAddress address;
+     MemoryContext oldContext;
+     
+     /* This makes it temporary table */
+     relation->relpersistence = RELPERSISTENCE_TEMP;
+     
+     /* Set up the CreateStmt */
+     stmt->relation = relation;
+     stmt->tableElts = create_xml_handle_columns();
+     stmt->constraints = NIL;
+     stmt->inhRelations = NIL;  
+     stmt->partspec = NULL;
+     stmt->ofTypename = NULL; 
+     stmt->oncommit = ONCOMMIT_PRESERVE_ROWS; 
+     stmt->tablespacename = NULL;
+     stmt->if_not_exists = false;  
+     stmt->options = NIL;
+     stmt->accessMethod = NULL;
+     
+     GetUserIdAndSecContext(&save_userid, &save_sec_context);
+ 
+     PG_TRY();
+     {
+		/* Set current user to bbf_role_admin for create permissions.*/
+		sql_dialect = SQL_DIALECT_TSQL;
+		SetUserIdAndSecContext(get_bbf_role_admin_oid(), save_sec_context | SECURITY_LOCAL_USERID_CHANGE);
+		address = DefineRelation(stmt, RELKIND_RELATION, InvalidOid, NULL, NULL);
```

Comment:
instead of elevating the privilege to get_bbf_role_admin_oid and running this DefineRelation, can we pass ownerId=get_bbf_role_admin_oid (3rd arg)?

Reply:
similar to above comment, but what if there are errors? what if the table is created but an error is thrown and the metadata hangs around. then all future XML calls might fail because the relation name is not unique.